File: basic/cs205_c_cpp_2020s/lab14/rusrc/lab14_01.rs
mod lab14_01 {
    fn max<T: std::cmp::PartialOrd>(one: T, two: T) -> T {
        let biggest: T;
        if one < two {
            biggest = two;
        } else {
            biggest = one;
        }
        return biggest;
    }

    pub(crate) fn main() {
        let i_one = 3i64;
        let i_two = 5i64;
        println!("the max of {} and {} is {}", i_one, i_two, max(i_one, i_two));
        let d_one = 5.6f64;
        let d_two = 7.3f64;
        println!("the max of {} and {} is {}", d_one, d_two, max(d_one, d_two));
        let s_one = "donkey";
        let s_two = "apple";
        println!("the max of {} and {} is {}", s_one, s_two, max(s_one, s_two));
    }
}

fn main() {
    lab14_01::main();
}

File: basic/cs205_c_cpp_2020s/lab14/exercise/lab14_01.cpp
// SPDX-License-Identifier: AGPL-3.0-or-later
/*  CS205_C_CPP
    Copyright (C) 2020-2023 nanoseeds
*/
/**
 * @Github: https://github.com/Certseeds
 * @Author: nanoseeds
 * @Date: 2020-05-23 11:01:42
 * @LastEditors  : nanoseeds
 */
#include <iostream>
#include <string>

using namespace std;

//Make a template out of the prototype
template<typename T>
T Max(T one, T two);

int main() {
    int i_one = 3, i_two = 5;
    cout << "The max of " << i_one << " and " << i_two << " is "
         << Max<int>(i_one, i_two) << endl;
//Test your template on float and string types
    double double_one = 5.6f, double_two = 7.3f;
    cout << "The max of " << double_one << " and " << double_two << " is "
         << Max<double>(double_one, double_two) << endl;
    string str_one = "donkey", str_two = "apple";
    cout << "The max of " << str_one << " and " << str_two << " is "
         << Max<string>(str_one, str_two) << endl;
    return 0;
}

template<typename T>
T Max(T one, T two) {
    T biggest;
    if (one < two) {
        biggest = two;
    } else {
        biggest = one;
    }
    return biggest;
}
